server/remote_cache/content_addressable_storage_server/content_addressable_storage_server.go

```diff
@@ -456,12 +458,52 @@ func makeTreeCacheActionResult(blob *rspb.ResourceName) ([]byte, error) {
 	return proto.Marshal(ar)
 }
 
-func (s *ContentAddressableStorageServer) cacheTreeNode(ctx context.Context, treeCachePointer *digest.ResourceName, treeCache *capb.TreeCache) error {
-	if !isComplete(treeCache.GetChildren()) {
+func (s *ContentAddressableStorageServer) cacheTreeNode(ctx context.Context, rootDir *capb.DirectoryWithDigest, treeCachePointer *digest.ResourceName, treeCache *capb.TreeCache) error {
+	rootCache, childCaches, err := splitTree(rootDir, treeCache)
+	// XXX: Should we instead proceed as if splitting is disabled?
+	if err != nil {
+		return nil
+	}
+
+	allChildren := rootCache.GetChildren()
+	for _, childCache := range childCaches {
+		allChildren = append(allChildren, childCache.GetChildren()...)
+	}
+	if !isComplete(allChildren) {
 		// incomplete tree cache error will be logged by `isComplete`.
 		return nil
 	}
-	buf, err := proto.Marshal(treeCache)
+
+	var childBytesWritten = 0
+	if len(childCaches) > 0 {
+		mu := &sync.Mutex{}
+		eg, egCtx := errgroup.WithContext(ctx)
+		for _, childCache := range childCaches {
+			eg.Go(func() error {
+				childBuf, err := proto.Marshal(childCache)
+				if err != nil {
+					return err
+				}
+				childBlob, err := makeTreeCacheDigest(treeCachePointer.GetDigestFunction(), childBuf)
+				if err != nil {
+					return err
+				}
+				if err := s.cache.Set(egCtx, childBlob.ToProto(), childBuf); err != nil {
```

Comment:
nit: wanna add a `if not s.cache.Contains()` checks before any CAS sets we do?

Reply:
Yup, done